message_view: Allow quoting selected message text, via hotkey and menu.

N-Shar-ma opened the pull request:
Whenever the user has text selected within a single message, and uses the hotkey to quote and reply, this selection will be quoted. In case of no selection or selection not within a single message, the entirety of the currently focused message will be quoted like before.

Similarly, when the user selects text within a single message, opens that message's actions menu, and clicks the "Quote and reply" option, the selected text will be quoted. In case of no selection or selection containing any other message/s, the entirety of the message whose menu was opened will be quoted like before.
    
When partially quoting a message, it is the markdown version of the selection that is quoted, hence preserving any formatting. Any other elements of the message, outside of the content, selected presumably by accident (like the timestamp or sender name), are ignored.

Fixes: #19712.

**Screenshots and screen captures:**
Text selected within a single message:
![image](https://github.com/zulip/zulip/assets/68962290/7e6bfd03-8861-4509-8319-c80f6d7e4485)

On clicking the Quote and reply menu option / using the `>` hotkey:
![image](https://github.com/zulip/zulip/assets/68962290/b35ab5ad-e628-4a86-9210-7492489c6913)

Preview of the quoted text:
![image](https://github.com/zulip/zulip/assets/68962290/28093440-5210-4f13-9f78-32b034e41ee3)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

karlstolley commented:
Nice work here, @N-Shar-ma!

Quoting most basic Markdown works well.

One thing I noticed is that if you try to copy one or more lines from a code block, that 1) the code-block markdown will be lost, and 2) so will any line breaks. Since it seems like at least in dev and on CZO, those both are preserved with copying and pasting, maybe there's away to achieve something similar her?

N-Shar-ma commented:
@karlstolley Thanks for testing this out!

I just tested and realised that losing the block formatting and new lines happens with copy pasting too (both on dev and CZO) so I'm pretty sure that's a pre-existing separate issue, that I'll try to investigate and fix soon.

N-Shar-ma commented:
@karlstolley Here's the fix for the partial code block pasting issue: #28798!

alya commented:
This is cool!

Would it be possible to have the tooltip aligned to the highlighted part of the message, or is that hard?

<img width="273" alt="Screenshot 2024-02-01 at 11 08 26 PM" src="https://github.com/zulip/zulip/assets/2090066/5ab8411c-f7ea-4efc-bfb6-51a342664a3e">

timabbott commented:
Hmm. That would be nice -- especially for a longer message, the bit you've highlighted may be quite a ways from the top-center of the message. It might require a bit of fiddling with Tippy configuration options to position it near the selection.

https://github.com/atomiks/tippyjs/issues/623 suggests that it is possible.

N-Shar-ma commented:
@timabbott Thanks for that issue link, it was helpful! I ended up using a more complex hack inspired by the 1 in that thread. Since the selection here isn't always plain text (often includes block elements like paragraphs) the popover isn't always horizontally centered but it's much better than before, and now shifts to the bottom or sides of the selection as per the space available.

After testing on both Chrome and Firefox I noticed Chrome always has just 1 range in a selection, but Firefox often has multiple ranges (for example, a code block's start always splits the range), so I changed the code to accommodate for that. For ensuring code blocks, lists, etc format correctly in markdown, I retain up to 2 ancestors for selections deep in the message (so that, for example, the `pre` and `code` elements are preserved). Also, I ironed out a few bugs, and cleaned up the code, adding comments for clarity.

I think this is ready for review!

N-Shar-ma commented:
popover in the middle of a message pointing up at the selection:
![image](https://github.com/zulip/zulip/assets/68962290/5da61989-bff6-4e1f-bb7c-f73c241c5bfe)

timabbott commented:
A corner case where the placement logic doesn't do well is when the timestamp is included in the selection:

![image](https://github.com/zulip/zulip/assets/2746074/3673d9f8-3694-4a01-9d84-c79e3fa61808)

I think you can narrow the selection before doing your placement logic and that'll do better. Might be worth playing with some other more complex syntaxes that might have a wide container inside, like polls/spoilers/code blocks.

timabbott commented on web/src/compose_reply.js:
```diff
@@ -149,7 +153,10 @@ export function quote_and_reply(opts) {
-    function replace_content(message) {
+    function replace_content(
+        message,
+        raw_content = opts.selected_message_content_raw ?? message.raw_content,
```
Could this line be outside the function definition? It feels like we should either be passing a required `raw_content` parameter that the callers pass as appropriate, or just computing this on the first line; I tend to prefer the "pass a real parameter" approach.

timabbott commented on web/src/compose_reply.js:
```diff
@@ -184,8 +191,191 @@ export function quote_and_reply(opts) {
+        temp_div.append(range.cloneContents());
+        return temp_div.innerHTML;
+    }
+    // Preserve the structure of the html upto 2 levels of ancestors.
```
Can you extend this comment to say you're doing that?

More generally I think all of these functions would benefit from more extensive comments explaining the algorithm, and what the parameters mean.

timabbott commented on web/src/compose_reply.js:
```diff
@@ -184,8 +191,191 @@ export function quote_and_reply(opts) {
+                range.getBoundingClientRect(),
+            );
+        }
+    }
```
The three cases here all feel very similar, in a way that suggests they could be deduplicated with the use of an intermediate variable for `html_to_convert`, maybe including an `else {continue;}` block for cases where none of these `if` statements were true.

N-Shar-ma replied:
Thanks for noticing this; initially all branches were very different, so I hadn't realised the duplication. :sweat_smile:

timabbott commented on web/src/compose_reply.js:
```diff
@@ -184,8 +191,191 @@ export function quote_and_reply(opts) {
+                `.focused-message-list [data-message-id="${message_id}"] .message_content`,
+            ),
+            selection_bounds,
+        );
```
Would it be cleaner to move the `create_popover_reference_in_message` call into `show_quote_and_reply_popover`, and just pass `message_id, selection_bounds, selected_message_content_raw` instead?

timabbott commented on web/src/compose_reply.js:
```diff
@@ -184,8 +191,191 @@ export function quote_and_reply(opts) {
+
+    $("body").on("mouseup", ".messagebox", () => {
+        // This gives the browser time to update the value returned by window.getSelection(),
+        // especially needed for cases where a click clears the previous selection.
```
I'd spell out a bit more. E.g. "This is necessary to avoid showing the popover for whatever was selected before the browser before our click."

timabbott commented:
This is pretty exciting; posted a few comments. I think once those are addressed, we should deploy on chat.zulip.org to get more broad testing, and think about whether we want to offer anything else in the little popover menu (E.g. a "Copy" option might feel natural).

N-Shar-ma commented:
> I think you can narrow the selection before doing your placement logic and that'll do better.

@timabbott Thanks for this suggestion! I created a function to get a range's intersection with an html element, and used that to narrow down any too broad range. So thanks to this the popover appears closer to the selection (noticeably closer vertically).

_However_, whenever there is a complete block style element within the selection, even if its content spans only a fraction of it's full width, the popover will be horizontally centered with respect to the full width of the block element, not the content width. That's why in your 2nd screenshot, the popover is in the center, since there are multiple paragraphs (block style elements) within the selection. That's just how `getBoundingClientRect()` works, and given that the html structure of the selection can be super nested and complex, it isn't really possible to get the "content width" of the selection. :frowning_face:

alya commented:
I tried testing out some corner cases. From what I can tell, the tooltip doesn't appear if I select part of a `/me` message.

alya commented:
If you select from a poll, you get very different results than if you use the "Quote and reply" menu option -- the current text, not the original `/poll` message. That's arguably more useful, so I'm mostly wondering whether we should change something about the menu option.

alya commented:
Is it expected that a block quote gets converted to a `>` quote when you use the tooltip to quote-and-reply?

alya commented:
Somewhat similarly, a ``` code block gets converted to an code span, losing the line breaks.

![Screenshot 2024-02-05 at 9 58 04 PM](https://github.com/zulip/zulip/assets/2090066/0e4db0e2-e127-413f-a2b7-389ef6c5a0ef)

N-Shar-ma commented:
@alya `/me` messages are tricky, yes, since they are displayed very differently to normal messages (mainly that the sender name and message content are 1 element, not separate). Also, what to paste when the sender name is selected partially is another open question, so I do think this is better left as a follow up.

For polls (and even todo messages for that matter, which quote even better and cleaner via selection), I think a follow up, making the menu option behave similarly (and cleaning up the polls output - The extraneous tip etc) would be feasible.

As for the blockquotes converting to `>` which happens on copy pasting too, that is because our syntax for quotes is not standard markdown. It renders fine though, so I haven't looked into trying to change that. I'm not sure it's worth it?

Lastly, thanks for catching the bug with the language-less code blocks; I've identified and fixed the logical loophole responsible in the code!

alya commented:
@timabbott I haven't re-tested, but I'd be fine with this PR getting deployed to CZO for broader feedback at this stage.

timabbott commented:
This has been running happily on chat.zulip.org for a few days. I think the implementation is solid, so I think it's reasonable to start working on final polishing work -- where we want to document it, any wording/adjustments, etc.

andersk commented:
Copying my feedback from [#design > Quote part of a message #19712](https://chat.zulip.org/#narrow/stream/101-design/topic/Quote.20part.20of.20a.20message.20.2319712/near/1738699):

- This should not trigger in spectator mode.
- I was going to speculate that this might interfere with the native text selection popup on mobile browsers. On Android that popup ordinarily provides options like **Copy**, **Share**, **Select all**, **Web search**, and **Read aloud**. However, when I went to test, I found that we’ve already broken that popup for some other reason.
- In alignment with the proposal at [#design > Explaining quote-and-reply](https://chat.zulip.org/#narrow/stream/101-design/topic/Explaining.20quote-and-reply) to rename **Quote and reply** to **Quote message**, this should also be renamed **Quote selection**.
- Is this function really important enough to merit its own popover? I would have thought it enough to alter the function of the `>` shortcut and the **⋮** → **Quote and reply**/**Quote message** menu item (which would rename itself to **Quote selection**) when part of the message is selected, as originally proposed in [#19712](https://github.com/zulip/zulip/pull/19712). Meanwhile, the popover might encourage even more unnecessary quotation when normal topic replies suffice.
- If you select something and click **⋮**, the selection popover incorrectly reappears, and moreover, it reappears above the message actions menu if their positions overlap.
- Double-clicking a word opens both the compose box and the selection popover. It now takes two `Esc` presses to dismiss them both.
- Double-clicking and dragging sometimes opens *two* selection popovers, one over the first word and one over the selection midpoint, depending on how I do it on my touchpad.

alya commented:
Works for me in manual testing!

timabbott commented on web/src/compose_reply.js:
```diff
@@ -129,10 +130,35 @@ export function reply_with_mention(opts) {
-    const message_id = opts.message_id || message_lists.current.selected_id();
+    message_id ??= message_lists.current.selected_id();
     const message = message_lists.current.get(message_id);
+    quote_content ??= message.raw_content;
```
I don't love the use of the `??=` operator here -- write a proper `if` block -- it feels hard to read. 

Also you may benefit from using more variable names in this block. For example, rather than calling `selection_within_message_id()` twice, just define a `const selection_message_id = ...`. So this could look something like the following, which very clearly lays out the 3 cases:

```
if (opts.message_id) {
     message_id = opts.message_id;
     quote_content = opts.quote_content;
} else {
    const selection_message_id = selection_within_message_id();
    if (selection_message_id !== undefined) {
         message_id = selection_message_id;
         quote_content = get_message_selection();
    } else {
         message = message_lists.current.selected_message();
         assert message !== undefined;
         quote_content = message.raw_content;
         message_id = message.id
    }

}


```

N-Shar-ma replied:
@timabbott Did some clean up, and added some comments too. If it's readable enough now, I think this should be ready for CZO

N-Shar-ma commented on web/src/compose_reply.js:
```diff
@@ -168,27 +194,109 @@ export function quote_and_reply(opts) {
+        // If the common ancestor is the message_content div or its child, we can quote
+        // this entire range at least.
+        if (range_common_ancestor.classList?.contains("message_content")) {
+            html_to_convert = extract_range_html(range);
```
Good point, but no, there is no case here where it's guaranteed that the entire message content is selected.

timabbott commented:
Just to update this thread from #design, the popover got moved to #29081.

timabbott commented:
This looks great, merged, thanks @N-Shar-ma! There's a documentation follow-up to https://zulip.com/help/quote-and-reply needed.

alya commented:
https://github.com/zulip/zulip/pull/29285
